x/ibc/testing/chain.go

```diff
@@ -332,8 +351,31 @@ func (chain *TestChain) CreateTMClient(counterparty *TestChain, clientID string)
 // UpdateTMClient will construct and execute a 07-tendermint MsgUpdateClient. The counterparty
 // client will be updated on the (target) chain.
 func (chain *TestChain) UpdateTMClient(counterparty *TestChain, clientID string) error {
+	// inject trusted fields into last header
+	header := counterparty.LastHeader
+	// Get TrustedHeight
+	trustedHeight := chain.GetClientState(clientID).GetLatestHeight()
+	var (
+		trustedVals *tmtypes.ValidatorSet
+		ok          bool
+	)
+	if trustedHeight == uint64(counterparty.LastHeader.Height) {
+		trustedVals = counterparty.Vals
+	} else {
+		// NOTE: We need to get validators from counterparty at height: trustedHeight+1
+		// since the last trusted validators for a header at height h
+		// is the NextValidators at h+1 committed to in header h by
+		// NextValidatorsHash
+		trustedVals, ok = counterparty.GetValsAtHeight(int64(trustedHeight + 1))
+		if !ok {
+			return sdkerrors.Wrapf(ibctmtypes.ErrInvalidHeaderHeight, "could not retrieve trusted validators at trustedHeight: %d", trustedHeight)
+		}
+	}
+	header.TrustedHeight = trustedHeight
+	header.TrustedValidators = trustedVals
```

Comment:
Note the added complexity here. This is the tradeoff with this feature. We remove storage on the clients, at the expense of a more complicated relayer `UpdateClient` workflow.

To update a clientA on chainB, a relayer must now:

1. Query the latestHeight of clientA on chainB (or a past stored height if we're doing a past update)
2. Query the NextValidators for that height on chainA
3. Inject those fields into the update Header and then submit to chainB

Thus it requires some preliminary back-and-forth querying between the chains before submitting a valid update

Reply:
can we document these changes somewhere? maybe add a TODO item on #6483?

Reply:
Does the signed header have the validator set? I think the relayer just needs to query for the signed header (which it already does) and just supply the correct fields

Reply:
sweet thanks for adding this!